allennlp/models/encoder_decoders/encoder_decoder.py

```diff
@@ -0,0 +1,320 @@
+        self._decoder_cell = LSTMCell(self._decoder_input_dim, self._decoder_output_dim)
+        self._output_projection_layer = Linear(self._decoder_output_dim, self._num_classes)
+
+    @overrides
+    def forward(self,  # type: ignore
+                source_tokens: Dict[str, torch.LongTensor],
+                target_tokens: Dict[str, torch.LongTensor] = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        """
+        Decoder logic for producing the entire target sequence.
+
+        Parameters
+        ----------
+        source_tokens : Dict[str, torch.LongTensor]
+           The output of ``TextField.as_array()`` applied on the source ``TextField``. This will be passed
+           through a ``TextFieldEmbedder`` and then through an encoder.
+        target_tokens : Dict[str, torch.LongTensor], optional (default = None)
+           Output of ``Textfield.as_array()`` applied on target ``TextField``. We assume that the arget tokens
+           are also represented as a ``TextField``.
+        """
+        # (batch_size, input_sequence_length, encoder_output_dim)
+        embedded_input = self._source_embedder(source_tokens)
+        source_mask = get_text_field_mask(source_tokens)
+        encoder_outputs = self._encoder(embedded_input, source_mask)
+        final_encoder_output = encoder_outputs[:, -1]  # (batch_size, encoder_output_dim)
+        if target_tokens:
+            targets = target_tokens["tokens"]
+            target_sequence_length = targets.size()[1]
+            # The last input from the target is either padding or the end symbol. Either way, we don't have to
+            # process it.
+            num_decoding_steps = target_sequence_length - 1
+        else:
+            num_decoding_steps = self._max_decoding_steps
+        # TODO (pradeep): Define a DecoderState class?
+        decoder_hidden = final_encoder_output
+        with torch.cuda.device_of(decoder_hidden):
+            decoder_context = Variable(torch.zeros(1, self._decoder_output_dim))
+        last_predictions = None
+        step_logits = []
+        step_probabilities = []
+        step_predictions = []
+        for timestep in range(num_decoding_steps):
+            if self.training and all(torch.rand(1) >= self._scheduled_sampling_ratio):
+                input_choices = targets[:, timestep]
+            else:
+                if timestep == 0:
+                    # For the first timestep, when we do not have targets, we input start symbols.
+                    # (batch_size,)
+                    with torch.cuda.device_of(final_encoder_output):
+                        input_choices = Variable(torch.LongTensor([self._start_index]).expand_as(
+                                final_encoder_output[:, 0]))
+                else:
+                    input_choices = last_predictions
+            decoder_input = self._prepare_decode_step_input(input_choices, decoder_hidden,
```

Comment:
What do you think of `_get_decoder_input()` as a function name?

Reply:
Hmm.. `decoder_input` does not distinguish the embedded vector from the indices (not that the current name does a much better job). Also, this is just for one timestep.